Mathlib/LinearAlgebra/Matrix/Ideal.lean

```diff
@@ -113,141 +114,220 @@ end Ideal
 Given an ideal $J ≤ Mₙ(R)$, we send it to $\{Nᵢⱼ ∣ ∃ N ∈ J\}$.
 -/
 @[simps]
-def equivMatricesOver (i j : n) : TwoSidedIdeal R ≃ TwoSidedIdeal (Matrix n n R) where
+def equivMatricesOver [Nonempty n] [DecidableEq n] :
+    TwoSidedIdeal R ≃ TwoSidedIdeal (Matrix n n R) where
   toFun I := I.matricesOver n
-  invFun J := TwoSidedIdeal.mk'
-    { N i j | N ∈ J }
-    ⟨0, J.zero_mem, rfl⟩
-    (by rintro _ _ ⟨x, hx, rfl⟩ ⟨y, hy, rfl⟩; exact ⟨x + y, J.add_mem hx hy, rfl⟩)
-    (by rintro _ ⟨x, hx, rfl⟩; exact ⟨-x, J.neg_mem hx, rfl⟩)
-    (by
-      classical
-      rintro x _ ⟨y, hy, rfl⟩
-      exact ⟨diagonal (fun _ ↦ x) * y, J.mul_mem_left _ _ hy, by simp⟩)
-    (by
-      classical
-      rintro _ y ⟨x, hx, rfl⟩
-      exact ⟨x * diagonal (fun _ ↦ y), J.mul_mem_right _ _ hx, by simp⟩)
-  right_inv J := SetLike.ext fun x ↦ by
-    classical
-    simp only [mem_mk', Set.mem_image, SetLike.mem_coe, mem_matricesOver]
-    constructor
-    · intro h
-      choose y hy1 hy2 using h
-      rw [matrix_eq_sum_stdBasisMatrix x]
-      refine sum_mem fun k _ ↦ sum_mem fun l _ ↦ ?_
-      suffices
-          stdBasisMatrix k l (x k l) =
-          stdBasisMatrix k i 1 * y k l * stdBasisMatrix j l 1 by
-        rw [this]
-        exact J.mul_mem_right _ _ (J.mul_mem_left _ _ <| hy1 _ _)
-      ext a b
-      by_cases hab : a = k ∧ b = l
-      · rcases hab with ⟨ha, hb⟩
-        subst ha hb
-        simp only [StdBasisMatrix.apply_same, StdBasisMatrix.mul_right_apply_same,
-          StdBasisMatrix.mul_left_apply_same, one_mul, mul_one]
-        rw [hy2 a b]
-      · conv_lhs =>
-          dsimp [stdBasisMatrix]
-          rw [if_neg (by tauto)]
-        rw [not_and_or] at hab
-        rcases hab with ha | hb
-        · rw [mul_assoc, StdBasisMatrix.mul_left_apply_of_ne (h := ha)]
-        · rw [StdBasisMatrix.mul_right_apply_of_ne (hbj := hb)]
-    · intro hx k l
-      refine ⟨stdBasisMatrix i k 1 * x * stdBasisMatrix l j 1,
-        J.mul_mem_right _ _ (J.mul_mem_left _ _ hx), ?_⟩
-      rw [StdBasisMatrix.mul_right_apply_same, StdBasisMatrix.mul_left_apply_same,
-        mul_one, one_mul]
-  left_inv I := SetLike.ext fun x ↦ by
-    simp only [mem_mk', Set.mem_image, SetLike.mem_coe, mem_matricesOver]
-    constructor
-    · intro h
-      choose y hy1 hy2 using h
-      exact hy2 ▸ hy1 _ _
-    · intro h
-      exact ⟨of fun _ _ => x, by simp [h], rfl⟩
+  invFun J := { ringCon := J.ringCon.ofMatrix }
+  right_inv _ := ringCon_injective <| RingCon.matrix_ofMatrix _
+  left_inv _ := ringCon_injective <| RingCon.ofMatrix_matrix _
+
+theorem coe_equivMatricesOver_symm_apply (I : TwoSidedIdeal (Matrix n n R)) (i j : n) :
+    equivMatricesOver.symm I = {N i j | N ∈ I} := by
+  ext r
+  constructor
+  · intro h
+    exact ⟨stdBasisMatrix i j r, by simpa using h i j, by simp⟩
+  · rintro ⟨n, hn, rfl⟩
+    intros i j'
+    simpa [mem_iff] using
+      mul_mem_right I _ (stdBasisMatrix _ _ 1) (mul_mem_left I (stdBasisMatrix _ _ 1) _ hn)
```

Comment:
This can be "golf"ed with the new RingCon lemma `coe_ofMatrix_eq_relationMap` but it's not really shorter.
```suggestion
  · rintro ⟨n, hn, rfl⟩
    rw [SetLike.mem_coe, mem_iff, equivMatricesOver_symm_apply_ringCon,
      RingCon.coe_ofMatrix_eq_relationMap i j]
    exact ⟨n, 0, (I.mem_iff n).mp hn, rfl, rfl⟩
```

Reply:
Do you prefer this spelling? If so, feel free to apply the suggestion.